File: module/service-grpc/service_maintenance.go
package main

import (
	"iter"

	"github.com/pantopic/wazero-grpc-server/sdk-go"
	"github.com/pantopic/wazero-grpc-server/sdk-go/codes"
	"github.com/pantopic/wazero-grpc-server/sdk-go/status"

	internal "github.com/pantopic/config-bus/module/service-grpc/internal"
)

var (
	statusResp = &internal.StatusResponse{
		Version:     "3.6.5",
		DbSize:      28672,
		DbSizeInUse: 28672,
		IsLearner:   false,
		Header:      &internal.ResponseHeader{},
	}
)

func serviceMaintenanceInit() {
	grpc_server.NewService(`etcdserverpb.Maintenance`).
		Unary(`Alarm`, maintenanceAlarm).
		Unary(`Status`, maintenanceStatus).
		Unary(`Defragment`, maintenanceDefragment).
		Unary(`Hash`, maintenanceHash).
		Unary(`HashKV`, maintenanceHashKV).
		ServerStream(`Snapshot`, maintenanceSnapshot).
		Unary(`MoveLeader`, maintenanceMoveLeader).
		Unary(`Downgrade`, maintenanceDowngrade)
}

func maintenanceAlarm(in []byte) (out []byte, err error) {
	return
}

func maintenanceStatus(in []byte) (out []byte, err error) {
	out, err = grpcError(kvShard().Read(append(in, QUERY_HEADER), false))
	if err != nil {
		return
	}
	err = statusResp.Header.UnmarshalVT(out)
	if err != nil {
		return []byte(err.Error()), status.New(codes.Unknown, err.Error()).Err()
	}
	statusResp.Leader = statusResp.Header.MemberId
	statusResp.RaftIndex = uint64(statusResp.Header.Revision)
	statusResp.RaftTerm = 1
	statusResp.RaftAppliedIndex = uint64(statusResp.Header.Revision)
	return statusResp.MarshalVT()
}

func maintenanceDefragment(in []byte) (out []byte, err error) {
	return
}

func maintenanceHash(in []byte) (out []byte, err error) {
	return
}

func maintenanceHashKV(in []byte) (out []byte, err error) {
	return
}

func maintenanceSnapshot(in []byte) (out iter.Seq[[]byte], err error) {
	out = func(yield func([]byte) bool) {}
	return
}

func maintenanceMoveLeader(in []byte) (out []byte, err error) {
	return
}

func maintenanceDowngrade(in []byte) (out []byte, err error) {
	return
}

File: module-zig/service-grpc/service_maintenance.zig
//! Mirrors module/service-grpc/service_maintenance.go

const std = @import("std");
const pb = @import("pb/etcdserverpb.pb.zig");

const errors = @import("error.zig");
const module = @import("module.zig");
const types = @import("types.zig");
const util = @import("util.zig");

var scratch: [16 * 1024]u8 = undefined;
var fba = std.heap.FixedBufferAllocator.init(&scratch);
var out: [16 * 1024]u8 = undefined;

pub fn alarm(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}

pub fn status(in: []const u8) anyerror!void {
    const r = util.kvShard().Read(util.withSuffix(in, types.QUERY_HEADER), false);
    if (r.err) |e| std.debug.panic("{s}", .{e});
    if (r.val != 1) {
        return module.server.sendErr(errors.grpcCode(r.res), r.res);
    }
    fba.reset();
    var reader = std.Io.Reader.fixed(r.res);
    const header = pb.ResponseHeader.decode(&reader, fba.allocator()) catch |err| {
        return module.server.sendErr(errors.code_unknown, @errorName(err));
    };
    const resp = pb.StatusResponse{
        .header = header,
        .version = "3.6.5",
        .dbSize = 28672,
        .dbSizeInUse = 28672,
        .isLearner = false,
        .leader = header.member_id,
        .raftIndex = @intCast(header.revision),
        .raftTerm = 1,
        .raftAppliedIndex = @intCast(header.revision),
    };
    var writer = std.Io.Writer.fixed(&out);
    try resp.encode(&writer, fba.allocator());
    return module.server.send(writer.buffered());
}

pub fn defragment(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}

pub fn hash(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}

pub fn hashKV(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}

pub fn snapshotOpen(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}

pub fn snapshotClose() anyerror!void {
    return module.server.send("");
}

pub fn moveLeader(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}

pub fn downgrade(in: []const u8) anyerror!void {
    _ = in;
    return module.server.send("");
}
